### PR TITLE
Add date input validators

### PR DESCRIPTION
**Pull request by roseline124**
## Feature Background
We can use custom DateInput with material-ui's DatePicker supporting this feature.
But, it causes additional codes and package and there might be people (like me) who prefers DateInput to DatePicker because of some reasons(UI and format bug, etc). so I add this date validators.

## Changes
- Add date, datetime validators `minDate`, `maxDate`
- Add minDate, maxDate to `validation.spec.ts`
- Add translated message (en, fr)
- Add example in `CreateEdit.md` document

## Check Lists
- [x] pass unit test
- [x] pass lint

### CODE REVIEW

File: packages/ra-language-french/src/index.ts [hunk 1]
```diff
@@ -138,6 +138,8 @@ const frenchMessages: TranslationMessages = {
             maxLength: 'Maximum %{max} caractères',
             minValue: 'Minimum %{min}',
             maxValue: 'Maximum %{max}',
+            minDate: 'Minimum %{min} date',
```
Comment (fzaninotto) [inline packages/ra-language-french/src/index.ts:141]:
```suggestion
            minDate: 'Au plus tôt %{min}',
```

File: packages/ra-language-french/src/index.ts [hunk 2]
```diff
@@ -138,6 +138,8 @@ const frenchMessages: TranslationMessages = {
...
+            maxDate: 'Maximum %{max} date',
```
Comment (fzaninotto) [inline packages/ra-language-french/src/index.ts:142]:
```suggestion
            maxDate: 'Au plus tard %{max}',
```

File: packages/ra-core/src/form/validate.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import lodashMemoize from 'lodash/memoize';
+import format from 'date-fns/format';
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.ts:2]: Browsers have an API for formatting dates (Intl), no need for an external library

File: packages/ra-language-french/src/index.ts (hunk 1 above)
Comment (roseline124) [reply packages/ra-language-french/src/index.ts:141]: Thanks!

File: packages/ra-core/src/form/validate.ts [hunk 4]
```diff
@@ -198,15 +198,18 @@ export const minDate = memoize(
     (min, message = 'ra.validation.minDate', formatString = 'YYYY-MM-DD') => (
         value,
         values
-    ) =>
-        !isEmpty(value) && new Date(value) < new Date(min)
+    ) => {
+        const inputValue = new Date(value);
+        inputValue.setSeconds(0);
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.ts:203]: What's the rationale of this?

**Comment by roseline124** [comment]
> I'm not sure I understand the difference with `minValue` and `maxValue`. Is it only the formatting of the `min` and `max` values in the error message?

I admit `minValue` and `maxValue` can cover this feature. But, I think `minDate`, `maxDate` validators are more readable and the error message is also important, too.

1. comparing code using minDate to one not using

I I added `nowDate.setSeconds(0)` to my feature.
Because DateTimeInput can't choose seconds, the seconds is set as `0` when selecting date in DateTimeInput.
So If I had set the baseDate as `Fri Jul 10 2020 17:47:57` and selected `Fri Jul 10 2020 17:47`, the validator will throw error even If these are the same date.

```typescript
<DateTimeInput
  source="advertisementEndAt"
  validate={[minDate(new Date())]}
/>
```

```typescript
<DateTimeInput
  source="advertisementEndAt"
  validate={[
    (value, values) => {
      const nowDate = new Date()
      nowDate.setSeconds(0)
      return minValue(Number(new Date()))(Number(value), values)
    },
  ]}
/>
```

2. Error message
- minDate: `Must be at least 2020-07-10`
- minValue: `Must be at least 1594371535226`

I'm okay if you close this PR.
Thank you for reviewing my code

File: packages/ra-core/src/form/validate.ts (hunk 4 above)
Comment (roseline124) [reply packages/ra-core/src/form/validate.ts:203]: It's written below!

File: packages/ra-core/src/form/validate.ts (hunk 3 above)
Comment (roseline124) [reply packages/ra-core/src/form/validate.ts:2]: > Hi, these validators look great 👍
> 
> Thanks for contributing 😊.
> 
> I've one remark: Can we can consider using another name for these validators?
> 
> * Minimal date means the input should contain a date after the specified one
> * Maximal date means the input should contain a date before the specified one
> 
> So maybe you can name these methods `afterDate` and `beforeDate`. What do you think?



> Hi, these validators look great 👍
> 
> Thanks for contributing 😊.
> 
> I've one remark: Can we can consider using another name for these validators?
> 
> * Minimal date means the input should contain a date after the specified one
> * Maximal date means the input should contain a date before the specified one
> 
> So maybe you can name these methods `afterDate` and `beforeDate`. What do you think?

I read your comment. Thank you for reviewing my code!
I implemented code to allow the base date to be included.
You means that it'll be better if the code excludes the base date? or just changes the validator name?

Comment (roseline124) [reply packages/ra-core/src/form/validate.ts]: > Browsers have an API for formatting dates (Intl), no need for an external library

Thank you! 
I used `new Intl.DateTimeFormat().format(date)` also changed format options (DateTimeFormatOptions).
The option can be a little long like this. I don't know this is good. 
```
const options = {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
};
```
It needs to know `DateTimeFormatOptions` and  readability doesn't look good. 
What do you think?

File: docs/CreateEdit.md [hunk 5]
```diff
@@ -946,6 +949,7 @@ const validateEmail = email();
 const validateAge = [number(), minValue(18)];
 const validateZipCode = regex(/^\d{5}$/, 'Must be a valid Zip Code');
 const validateSex = choices(['m', 'f'], 'Must be Male or Female');
+const validateDate = [minDate(new Date())];
```
Comment (JulienMattiussi) [inline docs/CreateEdit.md:952]: Should use maxDate here to make this exemple more consistant with a Birthdate

File: docs/CreateEdit.md [hunk 6]
```diff
@@ -919,6 +919,8 @@ Alternatively, you can specify a `validate` prop directly in `<Input>` component
 * `required(message)` if the field is mandatory,
 * `minValue(min, message)` to specify a minimum value for integers,
 * `maxValue(max, message)` to specify a maximum value for integers,
+* `minDate(min, message)` to specify a minimum value for date,
```
Comment (fzaninotto) [inline docs/CreateEdit.md:922]: this doc is now incomplete - you must add the formatting options parameter

File: docs/CreateEdit.md (hunk 5 above)
Comment (fzaninotto) [reply docs/CreateEdit.md:952]: you should also add formatting options to illustrate usage

File: packages/ra-core/src/form/validate.ts [hunk 7]
```diff
@@ -179,6 +179,81 @@ export const maxValue = memoize(
...
+ */
+export const minDate = memoize(
+    (min, message = 'ra.validation.minDate', formatOptions = {}) => (
+        value,
+        values
+    ) => {
+        const input = new Date(value);
+        input.setSeconds(0);
+
+        return !isEmpty(value) && input < new Date(min)
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.ts:204]: `min` should be a `Date`, why do you create a new Date object

File: packages/ra-core/src/form/validate.ts [hunk 8]
```diff
@@ -179,6 +179,81 @@ export const maxValue = memoize(
...
+ * @param {string|function} message
+ * @param {object} formatOptions
+ *
+ * @example
+ *
+ * const fooValidators = [minDate(new Date(), 'Should be after today')];
+ * <DateInput name="foo" validate={fooValidators} />
+ */
+export const minDate = memoize(
+    (min, message = 'ra.validation.minDate', formatOptions = {}) => (
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.ts:197]: please type the args using TypeScript

File: packages/ra-core/src/form/validate.ts [hunk 9]
```diff
@@ -179,6 +179,81 @@ export const maxValue = memoize(
...
+ * <DateInput name="foo" validate={fooValidators} />
+ */
+export const maxDate = memoize(
+    (max, message = 'ra.validation.maxDate', formatOptions = {}) => (
+        value,
+        values
+    ) => {
+        const input = new Date(value);
+        input.setSeconds(0);
+        return !isEmpty(value) && input > new Date(max)
```
Comment (fzaninotto) [inline packages/ra-core/src/form/validate.ts:241]: same, `max` should already be a `Date`

**Pull request closed without merging**